Content Understanding GA SDK for .NET

yungshinlintw opened the pull request:
Azure Content Understanding GA SDK for .NET PR.

This is a new SDK release, and we haven't released any .NET SDK based on this service before.



# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

yungshinlintw committed:
SDK-GEN: Generate SDK with tsp-client on CU GA TypeSpec

yungshinlintw committed:
SDK-FIX: Suppress specific warnings in project file for Azure.AI.ContentUnderstanding for dup names (AZC0034) and sevice API naming (AZC0031).

yungshinlintw committed:
MIGRATION: Migrate Project related files

yungshinlintw committed:
SDK-FIX: Update AnalyzeRequest1 initialization to handle null modelDeployments in ContentUnderstandingClient

yungshinlintw committed:
SAMPLE: Migrate Analyze* Samples

yungshinlintw committed:
SAMPLE: Update .gitignore to include sample output directories generated by sample execution

yungshinlintw committed:
SDK-EXT: ContentField.Value

yungshinlintw committed:
SDK-EXT: Extension methods for ArrayField and ObjectField to enhance usability with Count property and indexers for accessing elements.

yungshinlintw committed:
SDK-EXT: Add extension methods for DocumentContent to retrieve fields by name using indexers.

yungshinlintw committed:
SDK-FIX: Enhance deserialization logic in ContentAnalyzer and SupportedModels to handle cases where service returns arrays instead of objects.

yungshinlintw committed:
SDK-EXT: Add extension methods for ContentUnderstandingClient to update analyzer properties synchronously and asynchronously.

yungshinlintw committed:
SAMPLE: Add new samples for CreateAnalyzer, AnalyzeUrlPrebuiltInvoice, DeleteAnalyzer, ListAnalyzers, and UpdateAnalyzer demonstrating various functionalities of the Content Understanding service.

yungshinlintw committed:
SDK-FIX: Improve serialization logic in AnalyzeRequest1 and AnalyzeResult to ensure proper handling of modelDeployments and response structure.

yungshinlintw committed:
SDK-FIX: Update deserialization logic in AudioVisualContent to handle both "keyFrameTimesMs" and "KeyFrameTimesMs" for improved compatibility with service responses.

yungshinlintw committed:
SDK-EXT: Add OperationExtensions class with GetOperationId method to enhance usability by extracting operation IDs from Operation-Location headers.

yungshinlintw committed:
SAMPLE: Add GetResultFile sample demonstrating how to retrieve keyframe images from video analysis, including setup instructions and error handling.

Commit:
Add comprehensive tests for Azure.AI.ContentUnderstanding

- Added new test projects to the solution file, including
  `Azure.AI.ContentUnderstanding.Tests` and `Azure.Core.TestFramework`.
- Introduced `Directory.Build.props` for shared project properties.
- Created test suites for binary and URL-based document analysis,
  custom analyzer creation, and CRUD operations.
- Implemented utility classes (`TestHelpers`, `ContentUnderstandingTestBase`)
  for common test functionality and sanitization.
- Added tests for field extraction, confidence validation, and
  analyzer lifecycle operations (create, update, delete, list).
- Included sample files (`mixed_financial_docs.pdf`, `sample_invoice.pdf`)
  for testing document analysis.
- Exposed internal members to testing frameworks via `InternalsVisibleTo`.

Commit:
Include shared source file in test project

Added a new `<ItemGroup>` to `Azure.AI.ContentUnderstanding.Tests.csproj` to include the shared source file `AzureResourceProviderNamespaceAttribute.cs`. The file is referenced from the `AzureCoreSharedSources` directory and linked to the `Shared\Core\AzureResourceProviderNamespaceAttribute.cs` path. This enables reuse of the shared source file without duplication.

yungshinlintw committed:
SERVICE-FIX: Allow 201/202 in copy. SDK-FIX: copy request

yungshinlintw committed:
SAMPLE: Add new samples for Classifier, AnalyzeBinaryFeatures, AnalyzeCategory, and UpdateDefaults demonstrating various functionalities of the Content Understanding service, including classifier creation, binary feature analysis, and default model deployment configuration.

yungshinlintw committed:
Merge branch 'yslin/cu_ga_sdk_tsp_client' of https://github.com/Azure/azure-sdk-for-net into yslin/cu_ga_sdk_tsp_client

yungshinlintw committed:
SAMPLE-MD: Add AnalyzeBinary.md and also add sample test infra

yungshinlintw committed:
SAMPLE: Add Sample01_AnalyzeBinary demonstrating PDF analysis using prebuilt-documentSearch

- Introduced a new sample project for analyzing PDF files with the Content Understanding service.
- Added `Program.cs` to handle configuration, client creation, and analysis logic.
- Created `README.md` for setup instructions and prerequisites.
- Updated `.gitignore` to prevent committing user-specific `appsettings.json` files while allowing `appsettings.json.sample`.

yungshinlintw committed:
SAMPLE: Add Sample02_AnalyzeUrl demonstrating document analysis from a URL

- Introduced a new sample project for analyzing documents from publicly accessible URLs using the `prebuilt-documentSearch` analyzer.
- Added `Program.cs` to handle configuration, client creation, and analysis logic.
- Created `README.md` for setup instructions and prerequisites.
- Updated `Sample01_AnalyzeBinary.md` to include next steps for exploring additional samples.
- Updated `.gitignore` to prevent committing test recordings (SessionRecords).

yungshinlintw committed:
SAMPLE: Add Sample00_ConfigureDefaults demonstrating model deployment configuration

- Introduced a new sample project for configuring and retrieving default model deployment settings for Microsoft Foundry resources.
- Added `Program.cs` to handle configuration, client creation, and model deployment logic.
- Created `Sample00_ConfigureDefaults.md` for detailed documentation on model deployment configuration.
- Added `README.md` for setup instructions and prerequisites.
- Implemented tests for the sample to ensure functionality and correctness.

yungshinlintw committed:
SDK: Introduce customization for ContentUnderstandingClient to manage default model deployment settings

- Added a new partial class `ContentUnderstandingClient.Customizations.cs` to customize the generated methods, specifically making the `UpdateDefaults` protocol methods internal.
- Implemented convenience extension methods in `ContentUnderstandingClient.Extensions.cs` for updating default model deployment settings using a dictionary, providing a simpler public API.
- Updated documentation to clarify the usage of the new convenience methods and the internal nature of the protocol methods.

yungshinlintw committed:
SAMPLE: Add Sample03_AnalyzeInvoice demonstrating invoice analysis using prebuilt-invoice analyzer

- Introduced a new sample project for analyzing invoices from a URL using the `prebuilt-invoice` analyzer.
- Added `Program.cs` to handle configuration, client creation, and invoice analysis logic.
- Created `Sample03_AnalyzeInvoice.md` for detailed documentation on invoice analysis.
- Added `README.md` for setup instructions and prerequisites.
- Implemented tests for the sample to ensure functionality and correctness.

yungshinlintw committed:
SAMPLE: Add Sample04_CreateAnalyzer

yungshinlintw committed:
SAMPLE: Add multiple new samples for analyzer management and classification

- Introduced Sample05_CreateClassifier demonstrating how to create and use a classifier analyzer for document categorization.
- Added Sample06_GetAnalyzer to retrieve information about both prebuilt and custom analyzers.
- Created Sample07_ListAnalyzers to list all available analyzers in a Microsoft Foundry resource.
- Implemented Sample08_UpdateAnalyzer for updating existing custom analyzers, including their descriptions and tags.
- Added Sample09_DeleteAnalyzer to demonstrate how to delete a custom analyzer.
- Each sample includes detailed documentation and setup instructions in corresponding README.md files.
- Implemented tests for all new samples to ensure functionality and correctness.

yungshinlintw committed:
SAMPLE: Add new samples for analyzing documents and retrieving results

- Introduced Sample10_AnalyzeConfigs demonstrating how to extract features from documents using the `prebuilt-documentSearch` analyzer with configurations for formulas, layout, and OCR.
- Added Sample11_AnalyzeReturnRawJson to show how to access raw JSON responses from analysis operations for advanced scenarios.
- Created Sample12_GetResultFile to demonstrate retrieving result files, such as keyframe images, from video analysis operations using the `GetResultFile` API.
- Each sample includes detailed documentation and setup instructions in corresponding README.md files.
- Implemented tests for all new samples to ensure functionality and correctness.

yungshinlintw committed:
SAMPLE: Add new samples for deleting analysis results and copying analyzers

- Introduced Sample13_DeleteResult demonstrating how to delete analysis results using the `DeleteResult` API, including detailed documentation and setup instructions.
- Added Sample14_CopyAnalyzer to illustrate how to copy an analyzer within the same resource using the `CopyAnalyzer` API, with comprehensive examples and explanations.
- Each sample includes corresponding README.md files for setup and prerequisites.
- Implemented tests for both new samples to ensure functionality and correctness.

yungshinlintw committed:
SAMPLE: Remove deprecated samples for document analysis and management

- Deleted multiple sample projects including `Sample01_AnalyzeBinary`, `Sample02_AnalyzeUrl`, `Sample03_AnalyzeInvoice`, and others that are no longer needed.
- Removed associated `Program.cs` and `.csproj` files to clean up the codebase.
- This cleanup helps streamline the sample offerings and focuses on more relevant and updated examples for users.

yungshinlintw committed:
SAMPLE: Update appsettings.json.sample and remove outdated UpdateAnalyzer and UpdateDefaults samples

- Updated `appsettings.json.sample` to reflect new endpoint formats and added new configuration options for model deployments.
- Removed outdated sample projects `UpdateAnalyzer` and `UpdateDefaults`, including their associated `Program.cs` and `.csproj` files, to streamline the sample offerings.
- This cleanup focuses on maintaining relevant and up-to-date examples for users, enhancing the overall usability of the SDK.

yungshinlintw committed:
TEST: Add AssemblyInfo and include AzureResourceProviderNamespaceAttribute

- Introduced a new `AssemblyInfo.cs` file to define the Azure resource provider namespace for the Content Understanding SDK.
- Updated the project file to include the `AzureResourceProviderNamespaceAttribute` from shared sources, enhancing the SDK's integration with Azure services.
- This addition supports better resource management and organization within the Azure ecosystem.

yungshinlintw committed:
SAMPLE-TEST: Remove AnalyzeBinaryTest and enhance Sample01_AnalyzeBinary with assertions

- Deleted the `AnalyzeBinaryTest.cs` file as part of the cleanup process to streamline the test suite.
- Enhanced `Sample01_AnalyzeBinary.cs` by adding assertions to validate the analysis results, ensuring that the sample demonstrates expected behaviors and outcomes.
- This update improves the reliability of the sample and aligns it with best practices for testing and validation in the SDK.

Commit:
Enhance tests for Content Understanding SDK

- Added assertions to Sample05_CreateClassifier to verify classifier creation and properties.
- Improved Sample06_GetAnalyzer with assertions for prebuilt analyzers and invoice analyzer validation.
- Enhanced Sample07_ListAnalyzers to assert analyzer properties and verify prebuilt analyzers.
- Updated Sample08_UpdateAnalyzer to include assertions for updated analyzer properties and tags.
- Added assertions in Sample09_DeleteAnalyzer to verify successful deletion of analyzers.
- Enhanced Sample10_AnalyzeConfigs with assertions for analysis results and extracted content.
- Improved Sample11_AnalyzeReturnRawJson to validate raw JSON response and structure.
- Added assertions in Sample12_GetResultFile to verify result file retrieval and content.
- Enhanced Sample13_DeleteResult with assertions for analysis completion and result deletion.
- Improved Sample14_CopyAnalyzer to verify properties of copied analyzers and updates.

Commit:
Remove obsolete test helper and update analyzer tests for improved functionality and clarity

yungshinlintw committed:
Restore the deleted test csproj, and code cleanup

yungshinlintw committed:
TEST: Clean up

yungshinlintw committed:
SAMPLE: Add runnable GrantCopyAuth

yungshinlintw committed:
README: Update Sample README.md

yungshinlintw committed:
SAMPLE: Delete duplicated samples

yungshinlintw committed:
SDK-GEN: Re-generate SDK and add customization for return AnalyzeResultOperation.

yungshinlintw committed:
SDK-EXT: Address OperationId timing issue

yungshinlintw committed:
SAMPLE: Update samples to use byte arrays directly instead of BinaryData for AnalyzeBinaryAsync and related methods

yungshinlintw committed:
SDK-EXT: Use BinaryData for AnalyzeBinary

yungshinlintw committed:
SAMPLE: Sample Update to use BinaryData for AnalyzeBinary

yungshinlintw committed:
SDK-EXT: Change AnalyzeResultOperation constructor access modifier to internal

yungshinlintw committed:
SDK-EXT: Update AnalyzeResultOperation to expose operation ID via Id property and adjust samples accordingly

yungshinlintw committed:
SDK-EXT: Remove redundant XML documentation for the AnalyzeResultOperation class

yungshinlintw committed:
SAMPLE: Update samples to use operation ID from Id property instead of OperationId

chienyuanchang committed:
Update Sample Readme (#54129)

* update summary

* add env variable reminder

* update Microsoft Foundry

* update Microsoft Foundry

---------

Co-authored-by: Chien Yuan Chang <chieny@waferwire.com>

yungshinlintw committed:
SDK-GEN: Update Content Understanding SDK to reflect changes in model definitions and serialization logic

- Updated `tsp-location.yaml` to point to the latest commit in the Azure REST API specs for ContentCateogryDefinition renaming.

yungshinlintw committed:
SDK-EXT: Remove suppression of AZC0034 warning in Content Understanding project file

yungshinlintw committed:
SDK-EXT: Add suppression for AZC0034 warnings in Suppression.cs to handle naming conflicts with FormRecognizer and DocumentIntelligence SDKs

yungshinlintw committed:
SDK-GEN: Update Content Understanding SDK to correct SupportedModels TypeSpec

yungshinlintw committed:
SDK-FIX: Fix known emitters (copyAnalyzer path, result) and serivce issues (KeyFramesTimesMs, copy returnings 201)

yungshinlintw committed:
SAMPLE: Sample update for ContentCategory renaming

yungshinlintw committed:
TEST: Enable GrantCopyAuth test

yungshinlintw committed:
TEST: Migrate recording to test asset repo

yungshinlintw committed:
API and README

yungshinlintw committed:
CI: Fix broken links and remove en-US in the link

yungshinlintw committed:
CI: Make sure that samples are not packable

yungshinlintw committed:
CI: Fix broken link and remove en-US in link

yungshinlintw committed:
CI: CHANGELOG: Update version to 1.0.0-beta.1 (Unreleased) and clarify initial release status

yungshinlintw committed:
CI: Update sample documentation links to point to the correct GitHub URLs and remove language-specific segments from Microsoft documentation links.

yungshinlintw committed:
CI: Update azure-cognitive-services

yungshinlintw committed:
CI: Address cspell issue

yungshinlintw committed:
CI: Address cspell.json issue

yungshinlintw committed:
CI: Updaet assets.json

yungshinlintw committed:
SAMPLE: Delete dupe line

yungshinlintw committed:
SDK-FIX: Re-implement the required CU  service workaround and emitter with customization.

yungshinlintw committed:
SDK-FIX: Fix CI issues due to blank lines

yungshinlintw committed:
CI: Update for sample code snippets consistency between MD and sample tests

yungshinlintw committed:
CI: Fix relative link to MD to be URL

yungshinlintw commented:
/azp run prepare-pipelines

yungshinlintw commented:
/azp run prepare-pipelines

Commit:
Enhance test assertions and logging for Content Understanding SDK samples

- Improved verification and logging in Sample13_DeleteResult.cs to ensure detailed analysis operation and deletion behavior checks.
- Added comprehensive assertions and logging in Sample14_CopyAnalyzer.cs to validate source analyzer creation, copying process, and properties preservation.
- Expanded assertions and logging in Sample15_GrantCopyAuth.cs to verify source analyzer creation for cross-resource copying and copy authorization details.

yungshinlintw committed:
Merge remote-tracking branch 'origin/main' into yslin/cu_ga_sdk_tsp_client

yungshinlintw committed:
SDK-GEN: Regenerate SDK after mergeing from main

jsquire reviewed (dismissed):
Please reach out to @ronniegeraghty to request a label for the library.   Once he's assigned one, you'll want to register at least two of your team in CODEOWNERS so that you're able to perform team review of PRs.

 Please review the Azure SDK onboarding documentation and use the associated Teams channel for support.

- [Azure SDK onboarding](https://eng.ms/docs/products/azure-developer-experience/onboard/requirements) _(Microsoft internal)_
- [Azure SDK onboarding assistance](https://teams.microsoft.com/l/channel/19%3ade3fce22c2994be18cac50502c55f717%40thread.skype/Azure%2520SDK%2520Onboarding?groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) _(Microsoft internal)_

chienyuanchang committed:
use mixed_financial_docs for classifier (#54175)

Co-authored-by: Chien Yuan Chang <chieny@waferwire.com>

yungshinlintw committed:
SAMPLE: Removed snippets not used in MD file

yungshinlintw committed:
Update assets.json for Sample05_CreateClassifier test recordings

yungshinlintw committed:
Update assets.json tag to include new Sample05_CreateClassifier recordings

yungshinlintw committed:
SAMPLE: Update classifier test sample file

yungshinlintw committed:
TEST: Update assets.json tag to reflect latest changes in the Content Understanding SDK

yungshinlintw committed:
CI: Update samples to include file path and document URL placeholders to make sure that they can compile

- Added placeholders for `<document_url>` in Sample04_CreateAnalyzer and Sample06_GetAnalyzer.
- Included `<file_path>` placeholders in Sample05_CreateClassifier for binary analysis.
- Updated Sample15_GrantCopyAuth to use explicit source and target resource identifiers and API keys.
- Enhanced clarity in sample code for better user understanding.

yungshinlintw committed:
CI: Fix MD file link

yungshinlintw committed:
PR: Remove local editconfig

Commit:
Merge branch 'yslin/cu_ga_sdk_tsp_client' of https://github.com/Azure/azure-sdk-for-net into changjian-wang/improve_tests_for_cu_ga_sdk_tsp_client

yungshinlintw committed:
PR: Feedback to add CODEOWNERS and better nuget description

Commit:
TEST: Enhance console output for analysis operation verification in Sample02_AnalyzeUrl

yungshinlintw commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample00_ConfigureDefaults/Sample00_ConfigureDefaults.csproj:
```diff
@@ -0,0 +1,33 @@
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
```
@jsquire  Thanks for the feedback.

I have followed the sample guildeline to create Sample*.md, and also recorded test for testing samples in tests/samples/Sample*.cs. In the tests, we have also added assertions to ensure that the result is expected.

The reason that I added these "stand-alone" samples is to give developers sample project that they can just compile and run, without the need to copy/paste code from the md files and create their own projects. I think that this is a shortcoming of the existing markdown-based samples. I am trying to provide both so that the user can get the benefit of md files (better explanation, rich md formatting ability to provide references, etc.). And they can also have runnable samples that they can use immediately.

With that said, I understand your point about standalone samples. Besides, given that we have released a SDK yet, we don't have a nuget for the user to install. 

For this PR, I will remove all of the standalone dirs, and only leave the sample MD files. Once we release a SDK, I will updated these runnable samples back to reference the nuget, plus put test around them. I still need to think about the better way to test them that is not just checking for exceptions. Will discuss you offline on this.

yungshinlintw committed:
PR: Remove runnable samples per PR feedback.

yungshinlintw committed:
PR: Move cspell.yaml from .github/ to CU project

yungshinlintw committed:
CI: Fixed broken md file links

yungshinlintw committed:
Merge remote-tracking branch 'origin/changjian-wang/improve_tests_for_cu_ga_sdk_tsp_client' into yslin/cu_ga_sdk_tsp_client

yungshinlintw committed:
SAMPLE: Update messages

yungshinlintw committed:
TEST: update assets.json for extended test cases

yungshinlintw committed:
CI: Fix incorrect cspell.json path

yungshinlintw committed:
TEST: Add client tests

yungshinlintw committed:
CI: Fix snippet issue

yungshinlintw committed:
TEST: Fix incorrect media file used in GetResultFile sample test

yungshinlintw committed:
TEST: Extend client test cases

yungshinlintw committed:
SAMPLE: Extend client test cases

yungshinlintw committed:
CI: Update sample snippet

yungshinlintw committed:
CI: Fix incorrect sample test cod

yungshinlintw committed:
SAMPLE: Remove asserstions in GrantCopyAuth sample

yungshinlintw committed:
CODEOWNERS: Fix formatting

jsquire commented on .github/CODEOWNERS:
```diff
@@ -217,11 +217,14 @@
+
 # PRLabel: %Cognitive - Form Recognizer
-/sdk/documentintelligence/                                         @vkurpad
+/sdk/documentintelligence/                                         @vkurpad @yungshinlintw @bojunehsu
```
- `vkurpad` is missing the public membership to the `Microsoft` organization and write permissions.

yungshinlintw committed:
CODEOWNERS: Remove alias that does not meet req

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/AnalyzeResult.Customizations.cs:
```diff
@@ -0,0 +1,46 @@
+        /// Converts a response to an AnalyzeResult using the LRO result path.
+        /// </summary>
+        /// <remarks>
+        /// SDK-FIX: Customized to handle service response format inconsistency. The service sometimes wraps AnalyzeResult
```
Would this show up in IntelliSense and public docs?  Recommend moving implementation notes to comments.

yungshinlintw replied:
Not required anymore. Removed

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/ObjectField.Extensions.cs:
```diff
@@ -0,0 +1,33 @@
+        /// </summary>
+        /// <param name="fieldName">The name of the field to retrieve.</param>
+        /// <returns>The field if found, or null if not found.</returns>
+        public ContentField this[string fieldName]
```
Typical Dictionary behavior is to throw if key is not found.  Current behavior is arguably easier to use though without introducing TryGetValue, Contains, ...  Thoughts?

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/DocumentContent.Extensions.cs:
```diff
@@ -0,0 +1,35 @@
+        /// </summary>
+        /// <param name="fieldName">The name of the field to retrieve.</param>
+        /// <returns>The field if found, or null if not found.</returns>
+        public ContentField this[string fieldName]
```
If the main difference is between `content["Total"]` vs. `content.Fields["Total"]`, I wonder if it is worthwhile introducing this syntactic shortcut.  Being more explicit arguably makes the code easier to read.

yungshinlintw replied:
Removed to be in favor of documentContent.Fields?.GetFieldOrDefault("Total")

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/ContentUnderstandingClient.Extensions.cs:
```diff
@@ -0,0 +1,110 @@
+        /// <exception cref="ArgumentException"> <paramref name="analyzerId"/> is an empty string, and was expected to be non-empty. </exception>
+        /// <exception cref="RequestFailedException"> Service returned a non-success status code. </exception>
+        /// <returns> The response returned from the service. </returns>
+        public static Response UpdateAnalyzer(this ContentUnderstandingClient client, string analyzerId, ContentAnalyzer resource, CancellationToken cancellationToken = default)
```
Why weren't convenience methods auto generated?

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/ContentUnderstandingClient.Extensions.cs:
```diff
@@ -0,0 +1,110 @@
+
+        /// <summary> Update default model deployment settings. </summary>
+        /// <remarks>
+        /// This is the recommended public API for updating default model deployment settings.
```
Suggest we remove this remark.  Protocol methods are already tagged in their summaries.

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/ContentUnderstandingClient.Customizations.cs:
```diff
@@ -0,0 +1,271 @@
+            return null;
+        }
+
+        // SDK-FIX: Response classifier to accept both 201 and 202 status codes (service inconsistently returns both)
```
Need to understand why.

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/AudioVisualContent.Customizations.cs:
```diff
@@ -0,0 +1,272 @@
+            }
+            if (Optional.IsCollectionDefined(KeyFrameTimesMs))
+            {
+                // SDK-FIX: Serialize as "KeyFrameTimesMs" (capital K) to match service response format instead of "keyFrameTimesMs"
```
Though this method is likely never used, I recommend that we write using "keyFrameTimesMs" so that we don't need additional changes once service is fixed.

yungshinlintw replied:
Removed

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/README.md:
```diff
@@ -0,0 +1,260 @@
+# Azure ContentUnderstanding client library for .NET
```
```suggestion
# Azure Content Understanding client library for .NET
```

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/Generated/ContentUnderstandingClient.cs:
```diff
@@ -0,0 +1,1296 @@
+            }
+        }
+
+        /// <summary> Extract content and fields from input. </summary>
```
Why is this not marked as Protocol Method?

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/Generated/ContentUnderstandingClient.cs:
```diff
@@ -0,0 +1,1296 @@
+            Argument.AssertNotNull(resource, nameof(resource));
+
+            Operation<BinaryData> result = await CreateAnalyzerAsync(waitUntil, analyzerId, resource, allowReplace, cancellationToken.ToRequestContext()).ConfigureAwait(false);
+            return ProtocolOperationHelpers.Convert(result, response => ContentAnalyzer.FromLroResponse(response), ClientDiagnostics, "ContentUnderstandingClient.CreateAnalyzerAsync");
```
Emitter bug.  LRO resource creation with [PUT](https://github.com/microsoft/api-guidelines/blob/vNext/azure/ConsiderationsForServiceDesign.md#create-or-replace-a-resource-requiring-additional-long-running-processing) returns the resource JSON without status monitor wrapper.  FromLroResponse() should not expect `result`.

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/AnalyzeResult.Customizations.cs:
```diff
@@ -0,0 +1,46 @@
+        /// </summary>
+        /// <remarks>
+        /// SDK-FIX: Customized to handle service response format inconsistency. The service sometimes wraps AnalyzeResult
+        /// in a "result" property, and sometimes returns it directly. This workaround uses TryGetProperty to handle both formats.
```
We don't need this for AnalyzeResult.

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/README.md:
```diff
@@ -0,0 +1,260 @@
+* **`prebuilt-audioSearch`** - Transcribes audio content with speaker diarization, timing information, and conversation summaries. Supports multilingual transcription.
+* **`prebuilt-videoSearch`** - Analyzes video content with visual frame extraction, audio transcription, and structured summaries. Provides temporal alignment of visual and audio content.
+
+> **Note:** The prebuilt analyzers use the `prebuilt-{type}Search` naming pattern (not `prebuilt-{type}Analyzer`). This is a recent change in the Content Understanding service.
```
Given this is the first release of SDK, I wonder if talking about prebuilt-*Analyzer just adds to confusion.

bojunehsu replied:
Recommend using prebuilt-{modality}Search.  type is too vague.

yungshinlintw replied:
Remove the description for 'prebuilt-*Analzyer'

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/README.md:
```diff
@@ -0,0 +1,260 @@
+
+The API returns different content types based on the input:
+
+* **`document`** - For document files (PDF, images, Office documents). Contains pages, tables, figures, paragraphs, and markdown representation.
```
From an SDK level, perhaps we should use DocumentContent and AudioVideoContent instead of describing the enum value.

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/README.md:
```diff
@@ -0,0 +1,260 @@
+### Common Issues
+
+**Error: "Access denied due to invalid subscription key or wrong API endpoint"**
+- Verify your endpoint URL is correct and includes the trailing slash
```
Please verify that trailing slash is expected.

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample00_ConfigureDefaults.md:
```diff
@@ -0,0 +1,111 @@
+
+Content Understanding prebuilt analyzers require specific GPT model deployments to function:
+
+- **GPT-4.1** - Used by most prebuilt analyzers (e.g., `prebuilt-invoice`, `prebuilt-receipt`, `prebuilt-idDocument`)
```
Change all to use enum value expected by service.  `gpt-4.1`

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample01_AnalyzeBinary.md:
```diff
@@ -0,0 +1,187 @@
+
+To get started you'll need a **Microsoft Foundry resource**. See [README][README] for prerequisites and instructions.
+
+### ⚠️ IMPORTANT: Configure Model Deployments First
```
Do we need to emphasize this so much?

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample01_AnalyzeBinary.md:
```diff
@@ -0,0 +1,187 @@
+
+### Using API Key
+
+> **⚠️ Security Warning:** API key authentication is **not secure** for production use. API keys are sensitive credentials that should not be hardcoded or committed to source control. This method is **only recommended for testing purposes with test resources**. For production applications, use `DefaultAzureCredential` or other Azure AD-based authentication methods.
```
Can we unify the language across different doc pages?

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample01_AnalyzeBinary.md:
````diff
@@ -0,0 +1,187 @@
+}
+```
+
+The markdown output includes structured text with preserved formatting and hierarchy, table representations in markdown format, figure descriptions for images/charts/diagrams, and layout preservation maintaining document structure.
````
Rewrite for clarity.

bojunehsu commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample01_AnalyzeBinary.md:
```diff
@@ -0,0 +1,187 @@
+    Console.WriteLine($"Document type: {documentContent.MimeType ?? "(unknown)"}");
+    Console.WriteLine($"Start page: {documentContent.StartPageNumber}");
+    Console.WriteLine($"End page: {documentContent.EndPageNumber}");
+    Console.WriteLine($"Total pages: {documentContent.EndPageNumber - documentContent.StartPageNumber + 1}");
```
Remove

yungshinlintw committed:
README: PR feedback and casing for headers

yungshinlintw committed:
PR: Address PR feedback

yungshinlintw commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/README.md:
```diff
@@ -0,0 +1,260 @@
+
+#### Step 3: Configure Model Deployments (Required for Prebuilt Analyzers)
+
+> **IMPORTANT:** Before using prebuilt analyzers, you must configure the model deployments. This is a **one-time setup per Microsoft Foundry resource** that maps your deployed models to the prebuilt analyzers.
```
Removed the reference to UI today as it's not completely working yet. And add a pointer to sample 00 which shows how to update defaults.

yungshinlintw committed:
README: PR Feedback

yungshinlintw committed:
Sample MD: PR feedback

yungshinlintw committed:
Merge remote-tracking branch 'origin/main' into yslin/cu_ga_sdk_tsp_client

yungshinlintw committed:
SDK-GEN: Re-generate SDK with TypeSpec commit 5877ae086b6090761a921421ec6dc712baf3268b to reflect ContentCategoryDefinition change and also workaround with copyAnalyzer issue

yungshinlintw committed:
CUSTOMIZATION: Update customization to match the re-generated SDK

yungshinlintw committed:
SDK-GEN: Update TypeSpec commit to 4af871c7b6129cc1e2537259c9a73789e55c0703. SupportedModels.completion/embedding should be optional

yungshinlintw committed:
EMITTER-FIX: Udpate comments on CreateAnalyzer/CopyAnalyzer, and remove unnecessary customization for AnalyzeResult

yungshinlintw committed:
Merge remote-tracking branch 'origin/main' into yslin/cu_ga_sdk_tsp_client

yungshinlintw committed:
SDK-GEN: Re-generate with 970f986da5774177ebca954c926ad5dc059ae07b to pick up copyAnalyzer change

yungshinlintw committed:
Merge remote-tracking branch 'origin/main' into yslin/cu_ga_sdk_tsp_client

yungshinlintw committed:
SDK-GEN: Re-generate with commit 24fc7498d1c952e870fdc011a8ae3f71b626f7c1 to remove the need for ContentAnalyzer customization and also remove the need to check for 200/201 for copy

yungshinlintw committed:
SDK-EXT: Update field access based on API review
- Added ContentField.Fields.GetFieldOrDefault
- Have ObjectField index to throw KeyNotFound
- Update samples/tests
- Added unit tests for GetFieldOrDefault and ObjectField indexer
- ImprovedAnalyzeInvoice tests to assert all expected fields

yungshinlintw committed:
README and MD: PR feedback on casing. Renamed sample 00 to UpdateDefaults

yungshinlintw committed:
SAMPLE: Rename sample 00 to UpdateDefaults

yungshinlintw committed:
SDK-EXT: Refactor the way to extract operation Id

yungshinlintw committed:
TEST: Fix KnowledgeSource recording issue and fix expected results for invoice

yungshinlintw committed:
TEST: Upload new recording

yungshinlintw committed:
TEST: Improve DocumentProperties assertions

yungshinlintw commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/src/AudioVisualContent.Customizations.cs:
```diff
@@ -0,0 +1,272 @@
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        /// <param name="options">The client options for reading and writing models.</param>
+        protected override void JsonModelWriteCore(Utf8JsonWriter writer, ModelReaderWriterOptions options)
```
Removed

yungshinlintw committed:
SDK-EXT: Minor fixes for comments and unused code

yungshinlintw commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample01_AnalyzeBinary.md:
```diff
@@ -0,0 +1,187 @@
+
+For more information about the markdown format, see [Document Markdown][cu-document-markdown].
+
+## Access Document Properties with Type-Safe APIs
```
Updated

yungshinlintw committed:
SDK-GEN: Remove unneede serialization change for KeyFrameTimesMs

yungshinlintw committed:
SDK-EXT: Move UpdateDefaults/UpdateAnalyzer to be partial class. PR comments

yungshinlintw committed:
TEST: Update test recording

yungshinlintw commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample01_AnalyzeBinary.md:
```diff
@@ -0,0 +1,187 @@
+AnalyzeResultOperation operation = await client.AnalyzeBinaryAsync(
+    WaitUntil.Completed,
+    "prebuilt-documentSearch",
+    "application/pdf",
```
BinaryData has a MediaType but it does not do any detection, but it can be specified like:
```csharp
BinaryData.FromBytes(bytes, "application/pdf")
```

So we will use BinaryData.MeidaType is it is not null. Otherwise, default to applicaton/octet-stream

yungshinlintw committed:
SDK-EXT: Make contentType to be optional for AnalyzeBinary

yungshinlintw committed:
PR: Comments update

yungshinlintw committed:
SAMPLE: Add multi-modal modalities in analyze URL and PR feedback

yungshinlintw committed:
TEST: Update test recording

yungshinlintw committed:
CI: Fix README issues

yungshinlintw committed:
README: Fix CI issue and improve assersion on related sample tests

yungshinlintw committed:
README: remain gtp names

yungshinlintw commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/samples/Sample01_AnalyzeBinary.md:
```diff
@@ -0,0 +1,187 @@
+Content Understanding provides prebuilt analyzers that are ready to use without any configuration. These analyzers use the `*Search` naming pattern:
+
+- **`prebuilt-documentSearch`** - Extracts content from documents (PDF, images, Office documents) with layout preservation, table detection, figure analysis, and structured markdown output. Optimized for RAG scenarios.
+- **`prebuilt-imageSearch`** - Analyzes standalone images to generate descriptions, extract visual features, and identify objects and scenes within images. Optimized for image understanding and search scenarios. Note: Image analysis is not optimized for text extraction; use `prebuilt-documentSearch` for documents containing text.
```
Updated

yungshinlintw committed:
README: PR feedback

yungshinlintw committed:
README: Improve structure and consistencies

yungshinlintw committed:
TEST: Add Windows env setup

yungshinlintw committed:
README: Improve structure

yungshinlintw committed:
README: Improve text

yungshinlintw committed:
TEST: Remove test-resources.json

yungshinlintw committed:
SDK-EXT: Ensure ArrayField indexer throw ArgumentOutOfRangeException

yungshinlintw committed:
SDK-EXT: Comments update

yungshinlintw committed:
SDK-EXT: Comments update

yungshinlintw committed:
README: Improve sample 00

yungshinlintw committed:
SAMPLE: Improve 01 descriptions

yungshinlintw committed:
Merge remote-tracking branch 'origin/main' into yslin/cu_ga_sdk_tsp_client

yungshinlintw committed:
SDK-GEN: Update the latest SDK gen result after merging main

yungshinlintw committed:
SAMPLE: Improve sample 02

yungshinlintw committed:
SAMPLE: Improve structure

yungshinlintw committed:
SAMPLE: Improve sample 03

yungshinlintw committed:
SAMPLE: Improve sample 04

yungshinlintw committed:
SAMPLE: Improve Sample 04

yungshinlintw committed:
TEST: Update recording.

yungshinlintw committed:
SAMPLE: Improve sample 05

yungshinlintw committed:
SAMPLE: Fix 05 code

yungshinlintw committed:
SAMPLE: Update Sample 06 to display raw JSON responses for analyzers

yungshinlintw committed:
SAMPLE: Enhance Sample 06 to print analyzer properties and improve JSON output formatting

yungshinlintw committed:
SAMPLE: Update 06/07/08

yungshinlintw committed:
SAMPLE: Improve Sample 10 and 11

yungshinlintw committed:
SAMPLE: Update sample 12

yungshinlintw committed:
SAMPLE: Improve sample 12

yungshinlintw committed:
TEST: Push new recording

yungshinlintw committed:
SAMPLE: Improve sample 13 and 14

yungshinlintw committed:
SAMPLE: Fix CI issue

yungshinlintw committed:
SAMPLE: Update sample 15

yungshinlintw commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/tests/env.sample:
```diff
@@ -0,0 +1,111 @@
+# Azure AI Content Understanding Test Configuration
+# Copy this file to .env and fill in your actual values
```
This was used for local test only so that it's easier. However, I understand that this can be misleading for live testing. Currently, this SDK has used recorded test for all samples.

Will remove this file file and setup_est_env.cmd. And will look into add Test resource.

yungshinlintw committed:
PR: Remove local test env files

yungshinlintw committed:
SAMPLE: Add Models assignments

yungshinlintw committed:
SAMPLE: Change Update Analyzer to show update/add for now

yungshinlintw committed:
TEST: Update recording

yungshinlintw committed:
SAMPLE: Add references to CU Studio

yungshinlintw committed:
TEST: Fix CI issue

bojunehsu reviewed (approved):
Changes under src/ looks good.

yungshinlintw committed:
TEST: Added one case to ensure incremental update for UpdateDefautls and push new recordings

yungshinlintw committed:
TEST: Added test-resources

yungshinlintw commented:
/azp run prepare-pipelines

yungshinlintw commented on sdk/contentunderstanding/tests.yml:
```diff
@@ -0,0 +1,7 @@
+trigger: none
+
+extends:
```
Added test-resources.bicep to create required resources for running tests.

yungshinlintw committed:
Update assets.json with new recording tag for GrantCopyAuth test

yungshinlintw committed:
TEST: Update the capacity for GPT models

jsquire commented on sdk/contentunderstanding/Azure.AI.ContentUnderstanding/tests/Infrastructure/ContentUnderstandingClientTestEnvironment.cs:
```diff
@@ -0,0 +1,140 @@
+
+        // We are using assets from the Azure-Samples repository.
+        // Files are located at: https://github.com/Azure-Samples/azure-ai-content-understanding-dotnet/tree/main/ContentUnderstanding.Common/data
+        private const string FileUriFormat = "https://raw.githubusercontent.com/Azure-Samples/azure-ai-content-understanding-dotnet/main/ContentUnderstanding.Common/data/{0}";
```
You'd be safer using a commit hash rather than `main` here, as you don't control the sample and may fail if it changes out from underneath of you.

yungshinlintw replied:
Oh. I own the sample here as well, but I do agree that commit hash is safer. Will change in the next iteration

jsquire reviewed (approved):
Two minor comments, which are non-blocking.
